huaweicloud/services/rf/resource_huaweicloud_rf_stack.go

```diff
@@ -0,0 +1,403 @@
+		if err = deployStack(ctx, client, d, d.Timeout(schema.TimeoutCreate)); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	return resourceStackRead(ctx, d, meta)
+}
+
+func deployStack(ctx context.Context, client *golangsdk.ServiceClient, d *schema.ResourceData,
+	timeout time.Duration) error {
+	var (
+		stackId   = d.Id()
+		stackName = d.Get("name").(string)
+
+		opts = stacks.DeployOpts{
+			TemplateBody: d.Get("template_body").(string),
+			TemplateUri:  d.Get("template_uri").(string),
+			VarsBody:     d.Get("vars_body").(string),
+			VarsUri:      d.Get("vars_uri").(string),
+			StackId:      stackId,
+		}
+	)
+
+	deploymentId, err := stacks.Deploy(client, stackName, opts)
+	if err != nil {
+		return fmt.Errorf("error deploying stack resources: %s", err)
+	}
+	stateConf := &resource.StateChangeConf{
+		Pending: []string{"PENDING"},
+		Target:  []string{"COMPLETED"},
+		Refresh: stackStatusRefreshFunc(client, d.Id(), []string{
+			string(stacks.StackStatusDeploymentComplete),
+		}),
+		Timeout:      timeout,
+		Delay:        10 * time.Second,
+		PollInterval: 15 * time.Second,
+	}
+	resp, err := stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		if stack, ok := resp.(stacks.Stack); ok && stack.Status == string(stacks.StackStatusDeploymentFailed) {
+			return queryAllFailedEvents(client, stackId, stackName, deploymentId)
+		}
+	}
+	return err
+}
+
+func queryAllFailedEvents(client *golangsdk.ServiceClient, stackId, stackName,
+	deploymentId string) error {
+	opts := stacks.ListEventsOpts{
+		StackId:      stackId,
+		DeploymentId: deploymentId,
+	}
+	events, err := stacks.ListAllEvents(client, stackName, opts)
+	if err != nil {
+		return err
+	}
+
+	var mErr *multierror.Error
+	for _, event := range events {
+		if event.EventType == string(stacks.EventTypeError) {
+			mErr = multierror.Append(mErr, fmt.Errorf(event.EventMessage))
+		}
+	}
+	return mErr.ErrorOrNil()
+}
+
+// QueryStackById is a method to query stack details using its ID.
+func QueryStackById(client *golangsdk.ServiceClient, stackId string) (*stacks.Stack, error) {
+	resp, err := stacks.ListAll(client)
+	if err != nil {
+		return nil, err
+	}
+
+	filter := map[string]interface{}{
+		"ID": stackId,
+	}
+	result, err := utils.FilterSliceWithField(resp, filter)
+	if err != nil {
+		return nil, err
+	}
+	if len(result) < 1 {
+		return nil, golangsdk.ErrDefault404{}
+	}
+
+	stack, ok := result[0].(stacks.Stack)
+	if !ok {
+		return nil, fmt.Errorf("invaid object type, want 'stack.Stack', but '%T'", result[0])
+	}
+	log.Printf("[DEBUG] The details of the stack (%s) is: %#v", stackId, stack)
+
+	return &stack, nil
+}
+
+func stackStatusRefreshFunc(client *golangsdk.ServiceClient, stackId string, targets []string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := QueryStackById(client, stackId)
+		if err != nil {
+			if _, ok := err.(golangsdk.ErrDefault404); ok && len(targets) < 1 {
```

Comment:
do we need to raise this error if it's in a create/update process?

Reply:
In the create/update process, the length of target array is greater than 1.
And the error will be thrown.

Reply:
ok.